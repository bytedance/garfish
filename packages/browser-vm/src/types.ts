import { Sandbox } from './sandbox';
import type { LoaderOptions } from '@garfish/loader';

export type FakeWindow = Window & Record<PropertyKey, any>;
export type Module = (sandbox: Sandbox) => OverridesData | void;

export interface OverridesData {
  recover?: () => void;
  prepare?: () => void;
  created?: (context: Sandbox['global']) => void;
  override?: Record<PropertyKey, any>;
}

export interface ReplaceGlobalVariables {
  recoverList: Array<OverridesData['recover']>;
  prepareList: Array<OverridesData['prepare']>;
  createdList: Array<OverridesData['created']>;
  overrideList: Record<PropertyKey, any>;
}

export interface SandboxOptions {
  namespace: string;
  baseUrl?: string;
  fixBaseUrl?: boolean;
  fixStaticResourceBaseUrl?: boolean;
  disableWith?: boolean;
  strictIsolation?: boolean;
  disableElementtiming?: boolean;
  disableCollect?: boolean;
  modules?: Array<Module>;
  addSourceList?: (
    sourceInfo:
      | Array<{ tagName: string; url: string | URL | Request }>
      | { tagName: string; url: string | URL | Request },
  ) => void;
  loaderOptions?: LoaderOptions;
  styleScopeId?: () => string;
  el?: () => Element | ShadowRoot | null;
  protectVariable?: () => Array<PropertyKey>;
  insulationVariable?: () => Array<PropertyKey>;
}

export type StyledComponentCSSRulesData = Array<string>;
