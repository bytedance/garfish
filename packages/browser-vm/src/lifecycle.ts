import { SyncHook, PluginSystem } from '@garfish/hooks';
import type { interfaces } from '@garfish/core';
import type { FakeWindow } from './types';

export function sandboxLifecycle() {
  return new PluginSystem({
    closed: new SyncHook<[], void>(),
    stared: new SyncHook<[FakeWindow | undefined], void>(),
    appendNode: new SyncHook<[Element, Element, Element, string], void>(),
    beforeClearEffect: new SyncHook<[], void>(),
    afterClearEffect: new SyncHook<[], void>(),
    beforeInvoke: new SyncHook<
      [
        { code: string },
        string | undefined,
        Record<string, any>,
        interfaces.ExecScriptOptions | undefined,
      ],
      void
    >(),
    afterInvoke: new SyncHook<
      [
        { code: string },
        string | undefined,
        Record<string, any>,
        interfaces.ExecScriptOptions | undefined,
      ],
      void
    >(),
    invokeError: new SyncHook<
      [
        Error,
        string | undefined,
        Record<string, any> | undefined,
        interfaces.ExecScriptOptions | undefined,
      ],
      void
    >(),
  });
}
