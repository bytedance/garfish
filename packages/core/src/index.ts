export { Garfish } from './garfish';
export type { interfaces } from './interface';
