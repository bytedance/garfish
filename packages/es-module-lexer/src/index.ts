export * from '../dist/lexer';
export { ImportSpecifier, ExportSpecifier } from '../types/lexer.d';
