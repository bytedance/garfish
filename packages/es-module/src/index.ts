import { runtime } from './runtime';

console.log(runtime);
export function a() {}
