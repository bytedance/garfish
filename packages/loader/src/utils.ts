import {
  error,
  isPrimitive,
  isPlainObject,
  parseContentType,
} from '@garfish/utils';
import { Manager, Loader } from './index';

export async function request(url: string, config: RequestInit) {
  const result = await fetch(url, config || {});
  // Response codes greater than "400" are regarded as errors
  if (result.status >= 400) {
    error(`"${url}" load failed with status "${result.status}"`);
  }
  const code = await result.text();
  const type = result.headers.get('content-type');
  const mimeType = parseContentType(type);
  return { code, result, mimeType };
}

export function copyResult(result) {
  if (result.resourceManager) {
    result.resourceManager = (result.resourceManager as Manager).clone();
  }
  return result;
}

// Compatible with old api
export function mergeConfig(loader: Loader, url: string) {
  const extra = loader.requestConfig;
  const config = typeof extra === 'function' ? extra(url) : extra;
  return { mode: 'cors', ...config } as RequestInit;
}
