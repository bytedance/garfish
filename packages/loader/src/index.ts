import {
  SyncHook,
  SyncWaterfallHook,
  PluginSystem,
  AsyncHook,
} from '@garfish/hooks';
import {
  error,
  macroTask,
  __LOADER_FLAG__,
  isJsType,
  isCssType,
  isHtmlType,
} from '@garfish/utils';
import { StyleManager } from './managers/style';
import { ModuleManager } from './managers/module';
import { TemplateManager } from './managers/template';
import { JavaScriptManager } from './managers/javascript';
import { getRequest, copyResult, mergeConfig } from './utils';
import { FileTypes, cachedDataSet, AppCacheContainer } from './appCache';

// Export types and manager constructor
export * from './managers/style';
export * from './managers/module';
export * from './managers/template';
export * from './managers/javascript';

export type Manager =
  | StyleManager
  | ModuleManager
  | TemplateManager
  | JavaScriptManager;

export interface LoaderOptions {
  /**
   * The unit is byte
   */
  maxSize?: number;
}

export interface CacheValue<T extends Manager> {
  url: string;
  code: string;
  size: number;
  scope: string;
  fileType: FileTypes | '';
  resourceManager: T | null;
}

export interface LoadedHookArgs<T extends Manager> {
  result: Response;
  value: CacheValue<T>;
}

export enum CrossOriginCredentials {
  anonymous = 'same-origin',
  'use-credentials' = 'include',
}

type LifeCycle = Loader['hooks']['lifecycle'];

export type LoaderLifecycle = Partial<{
  [k in keyof LifeCycle]: Parameters<LifeCycle[k]['on']>[0];
}>;

export interface LoaderPlugin extends LoaderLifecycle {
  name: string;
  version?: string;
}

export class Loader {
  public personalId = __LOADER_FLAG__;
  public StyleManager = StyleManager;
  public ModuleManager = ModuleManager;
  public TemplateManager = TemplateManager;
  public JavaScriptManager = JavaScriptManager;
  /** @deprecated */
  public requestConfig: RequestInit | ((url: string) => RequestInit);

  public hooks = new PluginSystem({
    error: new SyncHook<[Error, { scope: string }], void>(),
    loaded: new SyncWaterfallHook<LoadedHookArgs<Manager>>('loaded'),
    clear: new SyncWaterfallHook<{
      scope: string;
      fileType?: FileTypes;
    }>('clear'),
    beforeLoad: new SyncWaterfallHook<{
      url: string;
      scope: string;
      requestConfig: ResponseInit;
    }>('beforeLoad'),
    fetch: new AsyncHook<[string, RequestInit], Response | void | false>(
      'fetch',
    ),
  });

  private options: LoaderOptions;
  private loadingList: Record<string, Record<string, Promise<CacheValue<any>>>>;
  private cacheStore: { [name: string]: AppCacheContainer };

  constructor(options?: LoaderOptions) {
    this.options = options || {};
    this.loadingList = Object.create(null);
    this.cacheStore = Object.create(null);
  }

  setOptions(options: Partial<LoaderOptions>) {
    this.options = { ...this.options, ...options };
  }

  clear(scope: string, fileType?: FileTypes) {
    const appCacheContainer = this.cacheStore[scope];
    if (appCacheContainer) {
      appCacheContainer.clear(fileType);
      this.hooks.lifecycle.clear.emit({ scope, fileType });
    }
  }

  clearAll(fileType?: FileTypes) {
    for (const scope in this.cacheStore) {
      this.clear(scope, fileType);
    }
  }

  usePlugin(options: LoaderPlugin) {
    this.hooks.usePlugin(options);
  }

  setLifeCycle(lifeCycle: Partial<LoaderLifecycle>) {
    this.hooks.usePlugin({
      name: 'loader-lifecycle',
      ...lifeCycle,
    });
  }

  loadModule(url: string) {
    return this.load<ModuleManager>({
      scope: 'modules',
      url,
      isRemoteModule: true,
    });
  }

  // Unable to know the final data type, so through "generics"
  async load<T extends Manager>({
    scope,
    url,
    isRemoteModule = false,
    crossOrigin = 'anonymous',
    defaultContentType = '',
  }: {
    scope: string;
    url: string;
    isRemoteModule?: boolean;
    crossOrigin?: NonNullable<HTMLScriptElement['crossOrigin']>;
    defaultContentType?: string;
  }): Promise<LoadedHookArgs<T>['value']> {
    const { options, loadingList, cacheStore } = this;

    const res = loadingList[url];
    // 增加 scope 作为缓存区分，相同 scope 的 loader 才做缓存
    // 否则可能出现，url 相同，不同子应用实例的共享同一 entryManager，在多个子应用同时 mount 时，在 vm sandbox 中会对 entryManager 的 document 进行赋值(app.entryManager.DOMApis.document = sandbox.global.document)
    // 这样当最后一个子应用实例对 entryManager.DOMApis.document 赋值时，其余子应用的 entryManager.DOMApis.document 将会指向最后一次实例化的子应用的 sandbox.global.document 变量，
    // dom 元素被创建时挂载的 sandbox id 为最后一个子应用的 sandbox id
    // 子应用通过 createElement 方法创建的 dom 元素将会被当做最后一个子应用的 deferClearEffects 副作用收集。
    // 当最后一个子应用被卸载时，其它子应用也将一起被卸载

    if (res && Object.keys(res).includes(scope)) {
      return macroTask(res[scope], 4);
    }

    let appCacheContainer = cacheStore[scope];
    if (!appCacheContainer) {
      appCacheContainer = cacheStore[scope] = new AppCacheContainer(
        options.maxSize,
      );
    }

    if (appCacheContainer.has(url)) {
      // When there is a real network request, some network io is consumed,
      // so we need to simulate the network io here to ensure that the timing of the request is correct when the cache timing is hit.
      return macroTask(copyResult(appCacheContainer.get(url)), 4);
    } else {
      // If other containers have cache
      for (const key in cacheStore) {
        const container = cacheStore[key];
        if (container !== appCacheContainer) {
          if (container.has(url)) {
            const result = container.get(url);
            cachedDataSet.add(result);
            appCacheContainer.set(url, result, result.fileType);
            return macroTask(copyResult(result), 4);
          }
        }
      }
    }

    const requestConfig = mergeConfig(this, url);
    // Tells browsers to include credentials in both same- and cross-origin requests, and always use any credentials sent back in responses.
    requestConfig.credentials = CrossOriginCredentials[crossOrigin];
    const resOpts = this.hooks.lifecycle.beforeLoad.emit({
      url,
      scope,
      requestConfig,
    });

    const request = getRequest(this.hooks.lifecycle.fetch);
    const loadRes = request(resOpts.url, resOpts.requestConfig)
      .then(({ code, size, result, type }) => {
        let managerCtor,
          fileType: FileTypes | '' = '';

        if (isRemoteModule) {
          fileType = FileTypes.module;
          managerCtor = this.ModuleManager;
        } else if (
          isHtmlType({ type, src: result.url }) ||
          isHtmlType({
            type: defaultContentType,
          })
        ) {
          fileType = FileTypes.template;
          managerCtor = this.TemplateManager;
        } else if (
          isJsType({ type: defaultContentType }) ||
          isJsType({ type, src: result.url })
        ) {
          fileType = FileTypes.js;
          managerCtor = this.JavaScriptManager;
        } else if (
          isCssType({ src: result.url, type }) ||
          isCssType({
            type: defaultContentType,
          })
        ) {
          fileType = FileTypes.css;
          managerCtor = this.StyleManager;
        }

        // Use result.url, resources may be redirected
        const resourceManager: Manager | null = managerCtor
          ? new managerCtor(code, result.url)
          : null;

        // The results will be cached this time.
        // So, you can transform the request result.
        const data = this.hooks.lifecycle.loaded.emit({
          result,
          value: {
            url,
            scope,
            resourceManager,
            fileType: fileType || '',
            // For performance reasons, take an approximation
            size: size || code.length,
            code: resourceManager ? '' : code,
          },
        });

        fileType && appCacheContainer.set(url, data.value, fileType);
        return copyResult(data.value as any);
      })
      .catch((e) => {
        __DEV__ && error(e);
        this.hooks.lifecycle.error.emit(e, { scope });
        throw e; // Let the upper application catch the error
      })
      .finally(() => {
        // 有 scope 区分之后无需置空
        // loadingList[url][scope] = null
      });

    loadingList[url]
      ? (loadingList[url][scope] = loadRes)
      : (loadingList[url] = { [scope]: loadRes });
    return loadRes;
  }
}
