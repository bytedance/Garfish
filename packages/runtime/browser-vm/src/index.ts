import { interfaces } from '@garfish/core';
import { warn, isPlainObject } from '@garfish/utils';
import { Sandbox } from './sandbox';
import { Module, SandboxOptions } from './types';
export { Sandbox } from './sandbox';

// export declare module
declare module '@garfish/core' {
  // This type declaration is used to extend garfish core
  export interface Garfish {
    getGlobalObject: () => Window & typeof globalThis;
    setGlobalValue(key: string, value?: any): void;
    clearEscapeEffect: (key: string, value?: any) => void;
  }

  export namespace interfaces {
    export interface Garfish {
      setGlobalValue(key: string, value?: any): void;
      getGlobalObject: () => Window & typeof globalThis;
      clearEscapeEffect: (key: string, value?: any) => void;
    }

    export interface SandboxConfig {
      modules?: Array<Module>;
    }

    export interface Config {
      protectVariable?: PropertyKey[];
      insulationVariable?: PropertyKey[];
      sandbox?: SandboxConfig;
    }

    export interface App {
      vmSandbox?: Sandbox;
    }

    export interface Plugin {
      openVm?: boolean;
    }
  }
}

// Strongly isolated webpack attributes
const webpackAttrs: PropertyKey[] = [
  'onerror',
  'webpackjsonp',
  '__REACT_ERROR_OVERLAY_GLOBAL_HOOK__',
];
if (__DEV__) {
  webpackAttrs.push('webpackHotUpdate');
}

// Compatible with old code
const compatibleOldModulesType = (config) => {
  if (isPlainObject(config.modules)) {
    __DEV__ && warn('"vm sandbox" modules should be an array');
    const list = [];
    for (const key in config.modules) {
      list.push(config.modules[key]);
    }
    config.modules = list;
  }
};

// Default export Garfish plugin
export default function BrowserVm() {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    // Garfish apis
    Garfish.getGlobalObject = () => {
      return Sandbox.getNativeWindow();
    };

    Garfish.setGlobalValue = (key, value) => {
      return (Garfish.getGlobalObject()[key] = value);
    };

    Garfish.clearEscapeEffect = (key, value?: any) => {
      const global = Garfish.getGlobalObject();
      if (key in global) global[key] = value;
    };

    // Use the default Garfish instance attributes
    let config: Partial<SandboxOptions> = { openSandbox: true };
    const options: interfaces.Plugin = {
      openVm: true,
      name: 'browser-vm',
      version: __VERSION__,

      afterLoad(appInfo, appInstance) {
        // Support for instance configuration, to ensure that old versions compatible
        const sandboxConfig = appInfo.sandbox || Garfish?.options?.sandbox;
        if (sandboxConfig === false) {
          config.openSandbox = false;
        }

        if (sandboxConfig) {
          config = {
            openSandbox:
              Sandbox.canSupport() &&
              sandboxConfig.open &&
              !sandboxConfig.snapshot,
            modules: sandboxConfig.modules || [],

            protectVariable: () => [
              ...(Garfish?.options?.protectVariable || []),
              ...(appInfo.protectVariable || []),
            ],
            insulationVariable: () => [
              ...(Garfish?.options?.insulationVariable || []),
              ...(appInfo.insulationVariable || []),
            ],
          };
        }
        options.openVm = config.openSandbox;

        if (!config.openSandbox) return;
        if (appInstance) {
          if (appInstance.vmSandbox) return;

          compatibleOldModulesType(config);
          // Create sandbox instance
          const sandbox = new Sandbox({
            openSandbox: true,
            namespace: appInfo.name,
            baseUrl: appInstance.entryManager.url,
            strictIsolation: appInstance.strictIsolation,
            sourceList: appInstance.sourceList,
            modules: [
              () => ({
                override: appInstance.getExecScriptEnv(false) || {},
              }),
              ...(config.modules || []),
            ],

            el: () => appInstance.htmlNode,
            protectVariable: config.protectVariable,
            insulationVariable: () => {
              return webpackAttrs.concat(config.insulationVariable() || []);
            },
          });

          appInstance.vmSandbox = sandbox;
          appInstance.global = sandbox.global;
          // Rewrite `app.execScript`
          appInstance.execScript = (code, env, url, options) => {
            sandbox.execScript(code, env, url, options);
          };
          // Use `Garfish.loader` instead of the `sandbox.loader`
          sandbox.loader = Garfish.loader;
        }
      },

      afterUnMount(appInfo, appInstance) {
        if (appInstance.vmSandbox) {
          // If the app is uninstalled,
          // the sandbox needs to clear all effects and then reset
          appInstance.vmSandbox.reset();
        }
      },

      afterMount(appInfo, appInstance) {
        if (appInstance.vmSandbox) {
          appInstance.vmSandbox.execScript(`
            if (typeof window.onload === 'function') {
              window.onload.call(window);
            }
          `);
        }
      },
    };

    return options;
  };
}
