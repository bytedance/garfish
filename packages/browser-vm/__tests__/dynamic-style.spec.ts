import { __MockHead__ } from '@garfish/utils';
import { Sandbox } from '../src/sandbox';
import { StyleManager } from '@garfish/loader';

describe('Sandbox: dynamic style', () => {
  let sandbox: Sandbox;

  const go = (code: string | (() => void)) => {
    if (typeof code === 'function') {
      code = `(${code})();`;
    }
    return `
      const sandbox = __debug_sandbox__;
      const Sandbox = sandbox.constructor;
      const nativeWindow = Sandbox.getNativeWindow();
      ${code}
    `;
  };

  // Mock style transformer
  const styleManagerProto = StyleManager.prototype;
  const originalTransformCode = styleManagerProto.transformCode;
  styleManagerProto.transformCode = function (code) {
    return originalTransformCode('.app ' + code);
  };

  beforeEach(() => {
    document.body.innerHTML = `<div id="root">123<div ${__MockHead__}></div></div>`;
    sandbox = new Sandbox({
      namespace: 'app',
      el: () => document.querySelector('#root'),
      baseUrl: 'http://test.app',
      modules: [
        () => ({
          recover() {},
          override: { go, jest, expect },
        }),
      ],
    });
  });

  it('set textContent', (done) => {
    sandbox.execScript(
      go(() => {
        const style = document.createElement('style');
        style.textContent = '.a { color: black; }';
        document.head.appendChild(style);
        expect([
          style.textContent,
          Array.from(style.sheet!.cssRules).map((x) => x.cssText),
        ]).toMatchSnapshot();
        done();
      }),
      { done },
    );
  });

  it('append text node', (done) => {
    sandbox.execScript(
      go(() => {
        const style = document.createElement('style');
        style.textContent = '.a { color: black; }';
        document.head.appendChild(style);
        style.appendChild(document.createTextNode('.b { color: white; }'));
        expect([
          style.textContent,
          Array.from(style.sheet!.cssRules).map((x) => x.cssText),
        ]).toMatchSnapshot();
        done();
      }),
      { done },
    );
  });

  it('insertRule', (done) => {
    sandbox.execScript(
      go(() => {
        const style = document.createElement('style');
        document.head.appendChild(style);
        style.sheet!.insertRule('.b { color: white; }');
        style.sheet!.insertRule('.a { color: black; }');
        style.sheet!.insertRule('.c { color: white; }', 2);
        expect(
          Array.from(style.sheet!.cssRules).map((x) => x.cssText),
        ).toMatchSnapshot();
        done();
      }),
      { done },
    );
  });
});
