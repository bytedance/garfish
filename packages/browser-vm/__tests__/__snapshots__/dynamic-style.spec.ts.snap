// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sandbox: dynamic style append text node 1`] = `
Array [
  ".app .a { color: black; }.app .b { color: white; }",
  Array [
    ".app .a {color: black;}",
    ".app .b {color: white;}",
  ],
]
`;

exports[`Sandbox: dynamic style insertRule 1`] = `
Array [
  ".app .a {color: black;}",
  ".app .b {color: white;}",
  ".app .c {color: white;}",
]
`;

exports[`Sandbox: dynamic style set textContent 1`] = `
Array [
  ".app .a { color: black; }",
  Array [
    ".app .a {color: black;}",
  ],
]
`;
